src/Util/Sniffs/FunctionDeclarations.php

```diff
@@ -0,0 +1,568 @@
+            if (isset($tokens[$stackPtr]['scope_opener']) === true) {
+                $scopeOpener = $tokens[$stackPtr]['scope_opener'];
+            }
+
+            $valid = [
+                T_STRING       => T_STRING,
+                T_CALLABLE     => T_CALLABLE,
+                T_SELF         => T_SELF,
+                T_PARENT       => T_PARENT,
+                T_NS_SEPARATOR => T_NS_SEPARATOR,
+            ];
+
+            for ($i = $tokens[$stackPtr]['parenthesis_closer']; $i < $phpcsFile->numTokens; $i++) {
+                if (($scopeOpener === null && $tokens[$i]['code'] === T_SEMICOLON)
+                    || ($scopeOpener !== null && $i === $scopeOpener)
+                ) {
+                    // End of function definition.
+                    break;
+                }
+
+                if ($tokens[$i]['code'] === T_NULLABLE) {
+                    $nullableReturnType = true;
+                }
+
+                if (isset($valid[$tokens[$i]['code']]) === true) {
+                    if ($returnTypeToken === false) {
+                        $returnTypeToken = $i;
+                    }
+
+                    $returnType .= $tokens[$i]['content'];
+                }
+            }
+
+            $end     = $phpcsFile->findNext([T_OPEN_CURLY_BRACKET, T_SEMICOLON], $tokens[$stackPtr]['parenthesis_closer']);
+            $hasBody = $tokens[$end]['code'] === T_OPEN_CURLY_BRACKET;
+        }//end if
+
+        if ($returnType !== '' && $nullableReturnType === true) {
+            $returnType = '?'.$returnType;
+        }
+
+        return [
+            'scope'                => $scope,
+            'scope_specified'      => $scopeSpecified,
+            'return_type'          => $returnType,
+            'return_type_token'    => $returnTypeToken,
+            'nullable_return_type' => $nullableReturnType,
+            'is_abstract'          => $isAbstract,
+            'is_final'             => $isFinal,
+            'is_static'            => $isStatic,
+            'has_body'             => $hasBody,
+        ];
+
+    }//end getProperties()
+
+
+    /**
+     * Checks if a given function is a PHP magic function.
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file where this token was found.
+     * @param int                         $stackPtr  The T_FUNCTION token to check.
+     *
+     * @return bool
+     */
+    public static function isMagicFunction(File $phpcsFile, $stackPtr)
+    {
+        if (Conditions::hasCondition($phpcsFile, $stackPtr, Tokens::$ooScopeTokens) === true) {
+            return false;
+        }
+
+        $name = $phpcsFile->getDeclarationName($stackPtr);
+        return self::isMagicFunctionName($name);
+
+    }//end isMagicFunction()
+
+
+    /**
+     * Verify if a given function name is the name of a PHP magic function.
+     *
+     * @param string $name The full function name.
+     *
+     * @return bool
+     */
+    public static function isMagicFunctionName($name)
```

Comment:
Actually, for the sniffs in PHPCS itself, the `is*Name()` methods are used more commonly than the `isMagicFunction()` etc methods.
See the implementation commits: 66cea8b697c584e96c49ee211f8c45cd7444c7cf, 6fb6c7aa1404ea9c70bbb6411dc4e03a04d0905e and 53a7c53e9b0bbd213a4e627aea4979588895194a

I've added them for convenience, to prevent adding duplicate code and to guarantee consistency in how this is checked, i.e.:
* For the `is*Name()` methods: making sure the name check is always done lowercased.
* For the `is*()` methods: making sure that not just the name, but also the scope in which the function exists is checked.

If the lowercasing for the name check wouldn't be necessary, I'd agree that the methods would be unnecessary overhead and sniffs and the `is*()` methods could just use `isset()`, however, as it is (PHP function names being case-insensitive), having the methods made sense, prevent adding duplicate code and make the (sniff) code more readable/descriptive.

As for the function call overhead, this has become so much less in PHP 7, that this IMO is not a consideration which could overrule code readability and consistency.

Reply:
Good points - makes sense. Thanks.